### DISCUSSION TITLE
Trouble getting zigpy_znp to install

### DISCUSSION THREAD

**Discussion by caylor93** [zigpy/zigpy-znp#134] [Q&A]
I'm trying to migrate from a Conbee II to a Nortek HUSBZB-1.   I have tried this on my production HA OS 7.2 and also on a clean install of the HA OS 6.6 both using a SD Card in a RPI-4B.

I have tried to install this following the directions located here:
https://github.com/zigpy/zigpy-znp/blob/dev/TOOLS.md#network-backup-beta

Also tried bellows following directions located here:
Both fail with the following error:  errors:

  error: command 'gcc' failed: No such file or directory
  ERROR: Failed building wheel for pycryptodome

Any thoughts on how to get past this is greatly appreciated.

**Comment by puddly** [comment]
You probably can't do this from within HA OS if it doesn't provide a C compiler. Set up a Python virtual environment on any other platform, there are installation instructions at the top of the documentation you linked to.

Migrating *to* a HUSBZB-1 is a little tricky because its IEEE address can only overwritten <ins>___one time___</ins> (unless you have access to a hardware debugger to completely re-flash it).  It's also beta quality so it *should* work but I haven't fully tested that exact codepath.

All of the new migration code is tracked in https://github.com/zigpy/zigpy-cli/pull/2. Follow the instructions in that pull request after setting up your Python environment to create a backup of your Conbee.

Open up the backup JSON file. If the network key frame counter is `0`, your Conbee is running old firmware and should be upgraded to the latest and the backup should be performed again.

Edit the backup JSON file and add the following `stack_specific` key to it, in order to replace the HUSBZB-1's IEEE address with the Conbee's:

```JSON
{
  "metadata": {
    ...
  },
  "stack_specific": {
    "ezsp": {
      "i_understand_i_can_update_eui64_only_once_and_i_still_want_to_do_it": true
    }
  },
  ...
```

Finally, write the backup JSON to your HUSBZB-1:

```console
$ zigpy -vv radio ezsp /dev/serial/by-id/... restore deconz.json
```

At this point, the Conbee *cannot* be used in proximity to the HUSBZB-1 because it would be forming an identical network. If you want to re-use the Conbee at your home, edit the backup JSON yet again and slightly change the IEEE address of the coordinator (e.g. increment the last octet by 1), the PAN ID, and the extended PAN ID. Then restore it to the Conbee.

Let me know if you have any questions or if something doesn't work. You're the first to try this migration so I'm hoping everything works 😄.

**Reply by caylor93** [reply]
Well, I assumed you could, because from that first link it says:

In Home Assistant OS
zigpy-znp is already installed by Home Assistant's ZHA component and is used to communicate with your radio. You only need to gain access to the correct Docker container and temporarily disable ZHA to run any of the tools:

Install [community "SSH & Web Terminal" addon](https://github.com/hassio-addons/addon-ssh/blob/main/ssh/DOCS.md) (not the official "Terminal & SSH" addon!) and configure your credentials. Note: the addon will fail to start if you use an insecure password so it's better to configure it with public key authentication.

Disable "Protection mode" in the addon's configuration.

At the shell, log into the main Home Assistant Docker container:

docker exec -it homeassistant bash
Since ZHA is still using your Zigbee radio, you must temporarily disable the ZHA integration before continuing. This is done with the three-dot dropdown menu in the "Integrations" section of the Home Assistant web UI.

**Reply by puddly** [reply]
You should be able to install the zigpy-cli packages from within HA OS, they do not require any new dependencies. You're probably trying to install the open coordinator backup branch of bellows, which is ancient and is likely trying to compile an old dependency from source.

**Reply by caylor93** [reply]
I tried both...followed the exact directions from that page and the command that I executed that created the error was:  pip install zigpy-znp

I believe this is not the bellows command right?  

I'll try the other method you indicate above...I'll use the Raspbian I have on another SD Card to try it.  I'm not a linux guy so hoping I can figure it all out....just may take a while.

**Reply by puddly** [reply]
I believe someone has set up https://github.com/zigpy/zigpy-cli/pull/2 on HA OS directly and it worked. Give that a try first.

**Reply by caylor93** [reply]
Maybe I'm doing something wrong.  Followed the "sample" directions there and getting the same errors when I did:

pip install \
    git+https://github.com/puddly/zigpy.git@puddly/new-radio-settings-api \
    git+https://github.com/puddly/zigpy-cli.git@puddly/zigpy-radio-api \
    git+https://github.com/puddly/zigpy-znp.git@puddly/new-radio-settings-api \
    git+https://github.com/puddly/zigpy-deconz.git@puddly/new-radio-api \
    git+https://github.com/puddly/bellows.git@puddly/zigpy-radio-api

  **error: command 'gcc' failed: No such file or directory**
  ----------------------------------------
**ERROR: Failed building wheel for pycryptodome**

Then when I try and run:
 zigpy -vv radio deconz /dev/serial/by-id/usb-dresden_elektronik_ingenieurtechnik_GmbH_ConBee_II_DE2193514-if00 backup deconz.json

It says:  **bash: zigpy: command not found**

I'll try it from Raspbian and see if I have any better luck.  

Was hoping this would work for me as I was trying to move from two separate dongles (one for Z-Wave and one for Zigbee) to the single Nortek one to clean things up a bit.  But not looking forward to manually migrating 60+ devices just to get to one dongle so may have to "wave the white flag" at that point and stick with two or suck it up and manually migrate.

**Reply by puddly** [reply]
Yeah if it can't compile a dependency then you can't do this from within HA OS. All of the packages are cross platform so try it in another Linux or macOS environment. Windows should work too.

**Reply by caylor93** [reply]
I think I'm closer.  It seems installed fine on Ubuntu 20.04.  I have the deconz.json file and the contents are:

(venv) ubuntu@ubuntu:~$ cat deconz.json
{
    "metadata": {
        "version": 1,
        "format": "zigpy/open-coordinator-backup",
        "source": "zigpy-cli@0.0.1",
        "internal": {
            "node": {
                "ieee": "00212effff05a153",
                "nwk": "0000",
                "type": "coordinator"
            },
            "network": {
                "tc_link_key": {
                    "key": "I removed the key from this post",
                    "frame_counter": 0
                },
                "tc_address": "00212effff05a153",
                "nwk_manager": "0000"
            },
            "link_key_seqs": {}
        }
    },
    "stack_specific": {},
    "coordinator_ieee": "00212effff05a153",
    "pan_id": "eacb",
    "extended_pan_id": "00212effff05a153",
    "nwk_update_id": 0,
    "security_level": 5,
    "channel": 15,
    "channel_mask": [
        15
    ],
    "network_key": {
        "key": "I removed the key from this post",
        "sequence_number": 0,
        "frame_counter": 17773475
    },
    "devices": []

I saw your note:
Open up the backup JSON file. If the network key frame counter is 0, your Conbee is running old firmware and should be upgraded to the latest and the backup should be performed again.

I upgraded to the latest firmware and re-ran it and it still shows "0"

So for the next step I need to go in where it says:   "stack_specific": {},

Replace that with:

  "stack_specific": {
    "ezsp": {
      "i_understand_i_can_update_eui64_only_once_and_i_still_want_to_do_it": true
    }

Do I replace "i_understand_i_can_update_eui64_only_once_and_i_still_want_to_do_it" with something?  What is it?

And then run the command:  zigpy -vv radio ezsp /dev/serial/by-id/... restore deconz.json

I have two different devices listed, any idea which I should use?  Is one the Zigbee and one the Zwave?
- usb-Silicon_Labs_HubZ_Smart_Home_Controller_71500ED6-if00-port0  
- usb-Silicon_Labs_HubZ_Smart_Home_Controller_71500ED6-if01-port0 

Thanks for your help on this

Kris

**Reply by Adminiuga** [reply]
> Migrating to a HUSBZB-1 is a little tricky because its IEEE address can only overwritten one time

Yep, unless you have 50 or more devices, it could be easier to re-pair everything. 
Or, if you could run two ha instances for a bit, setup nortek on the second instance, open network for joining (60s by default or 250 if you call the zha service directly)  and remove device from the 1st network. Most devices start looking for new network, if removed from an existing one. A bit easier than resetting each device. 
After all devices are re-paired, stop second instance, bring ZigBee.db to the old instance and reconfigure it for new stick.

**Reply by caylor93** [reply]
Thanks....sounds easier.

To remove the old Conbee from the first instance is that just pulling it from the RPI?  Or do I remove it from the ZHA integration?  Remove the ZHA integration itself?  

Also, I have automations that use the devices/entities, so once I move the  ZigBee.db am I going to have to rename each device and entities so that the automations work, or update automations to the new names/entities?  I guess that would apply to the lovelace cards too.

**Reply by caylor93** [reply]
Hello,  do you have any guidance on the question:

To remove the old Conbee from the first instance is that just pulling the dongle from the RPI? Or do I remove it from the ZHA integration? Remove the ZHA integration itself?

**Reply by Adminiuga** [reply]
Remove zha integration.

**Reply by Adminiuga** [reply]
But keep in mind, it will remove all the customization you made for zha entities.

**Reply by Hedda** [reply]
> I'm trying to migrate from a Conbee II to a Nortek HUSBZB-1.

@caylor93 Have you also upgraded firmware on the Nortek HUSBZB-1 before getting started? See tools in https://github.com/walthowd/husbzb-firmware and maybe even later versions from https://github.com/grobasoz/zigbee-firmware/tree/master/EM3581/%40walthowd

Also, while off-topic I'm wondering why you are migrating to Nortek HUSBZB-1 which is based on an old EM3581 chip that is today considered obsolete by Silicon Labs for new implementations. Though it should still work fine if use newer firmware than it ships with, it would be more future-proof to buy a Zigbee Coordinator with a newer chip. I can however not advise on a newer Silicon Labs based Zigbee Coordinator adapters as Silabs has suffered during the current worldwide chip shortage so availability is scarce (but I read that [Elelabs](https://elelabs.com/) is planning on releasing new models soon to replace their existing models).

**Reply by MattWestb** [reply]
deCONZ is not so "user friendly" but good hardware with payed "without prepared" support for end users.
HUSBZB-1 is old but its still running strong and i thing you cant finding one better power / reliever Zigbee chip and its also is having one OK ZW chip and its normally very stable and is not so tricky with interference from USB 3 and so on.

Its little as baying one Audi Q7 (2023 model) for chopping in the city then one A1 (2010 model) is more then enough and is much easier to parking.

**Reply by caylor93** [reply]
> why you are migrating to Nortek HUSBZB-1 which is based on an old EM3581 chip

I was trying to get down to a single dongle for Zigbee and ZWave.  Didn't look at the firmware or think to consider that the chip might be older.  I'm about 80% through a manual migration of both Zigbee and ZWave networks.  Gave up on trying to migrate the way puddly recommended.

**Reply by MattWestb** [reply]
The original Zigbee firmware is normally old and have some nasty bugs so upgrade it to EZSP 6.7.8.0 or test the newest if you like but its not tested so much and is not flagged stable.

Also if you is having good routers around you can blocking the coordinator having direct children so end devices must using routers and not getting problem then the coordinator is offline.
```
zha:
  zigpy_config:
    source_routing: true
    ezsp_config:
      CONFIG_MAX_END_DEVICE_CHILDREN: 0
```
PS: You must have least one router joined before you is doing this setting or you cant joining any devices at all.

**Reply by Hedda** [reply]
> I was trying to get down to a single dongle for Zigbee and ZWave. Didn't look at the firmware or think to consider that the chip might be older.

Ah, you are going by the old "_**a jack of all trades is a master of none, but oftentimes better than a master of one**_" saying. While that does sound like a good idea on paper it is generally a relatively bad idea in practice, especially when it comes to MCU radio adapters for IoT low-power communication protocols. At least the best practice today is to have one MCU radio USB adapter per protocol. 

I would not recommend Nortek HUSBZB-1 for a new purchase, however the old EM3581 chip used for Zigbee in it still keep up with the latest firmware can all Zigbee 3.0 devices released so far in the world, but the firmware is no longer getting new features and new functions so it will not be future-proof, however, it is easy to backup your Zigbee network to file and restore that to newer hardware with a newer Zigbee chip.

But you should know as well that Nortek HUSBZB-1 does also not have the latest Z-Wave chip. I only have a Z-Wave 500 series chip for its Z-Wave Controller, and the latest chips today is Z-Wave 700 series chips so again it is not future-proof. New devices using Z-Wave 700 series chips are backwards compatible with a Z-Wave 500 but you will not get all the benefits (like longer range and better battery management), however, it is easy to backup your Z-Wave network and restore that to new hardware with a newer Z-Wave 700 chip.

PS: If you are looking to try out more protocols in the future then keep in mind that will soon probably want a third MCU radio USB adapter for the new Matter over Thread protocol standard, (it will require a MCU radio USB adapter based on a new chip that supports OpenThread, which also happen to be the same type of newer MCU radio chips that can also support Zigbee 3.0 if just using a different firmware).
